packages/bff/src/services/authorizationService.ts

```diff
@@ -0,0 +1,152 @@
+) {
+  const readJwt = async (
+    identityToken: string,
+    logger: Logger
+  ): Promise<{
+    roles: string;
+    sessionClaims: SessionClaims;
+    selfcareId: string;
+  }> => {
+    const verified = await verifyJwtToken(identityToken, logger);
+    if (!verified) {
+      throw invalidJwtClaim("Token verification failed");
+    }
+
+    const decoded = decodeJwtToken(identityToken);
+
+    const {
+      success: validToken,
+      data: token,
+      error,
+    } = AuthToken.safeParse(decoded);
+    if (!validToken) {
+      const claim = error?.errors[0]?.path.join(".");
+      throw missingClaim(claim);
+    }
+
+    if (token.role) {
+      throw genericError("User roles in context are not in valid format");
+    }
+
+    const sessionClaims = SessionClaims.parse(token);
+
+    const selfcareId = token[ORGANIZATION].id;
+
+    return {
+      roles: token[USER_ROLES].join(","),
+      sessionClaims,
+      selfcareId,
+    };
+  };
+
+  const assertTenantAllowed = (selfcareId: string, origin: string): void => {
+    if (
+      !config.tenantAllowedOrigins.includes(origin) &&
+      !allowList.includes(selfcareId)
+    ) {
+      throw unknownTenantOrigin(selfcareId);
+    }
+  };
+
+  const buildJwtCustomClaims = (
+    roles: string,
+    tenantId: string,
+    selfcareId: string,
+    tenantOrigin: string,
+    tenantExternalId: string
+  ): CustomClaims => ({
+    [USER_ROLES]: roles,
+    [ORGANIZATION_ID_CLAIM]: tenantId,
+    [SELFCARE_ID_CLAIM]: selfcareId,
+    [ORGANIZATION_EXTERNAL_ID_CLAIM]: {
+      [ORGANIZATION_EXTERNAL_ID_ORIGIN_CLAIM]: tenantOrigin,
+      [ORGANIZATION_EXTERNAL_ID_VALUE_CLAIM]: tenantExternalId,
+    },
+  });
+
+  return {
+    getSessionToken: async (
+      correlationId: string,
+      identityToken: string,
+      logger: Logger
+    ): Promise<string> => {
+      const { sessionClaims, roles, selfcareId } = await readJwt(
+        identityToken,
+        logger
+      );
+
+      const { serialized } =
+        await interopTokenGenerator.generateInternalToken();
+
+      const headers = {
+        "X-Correlation-Id": correlationId,
+        Authorization: `Bearer ${serialized}`,
+      };
+
+      const tenantBySelfcareId =
+        await tenantProcessClient.selfcare.getTenantBySelfcareId({
+          params: { selfcareId },
+          headers,
+        });
+      const tenantId = tenantBySelfcareId.id;
+
+      const tenant = await tenantProcessClient.tenant.getTenant({
+        params: { id: tenantId },
+        headers,
+      });
+
+      assertTenantAllowed(selfcareId, tenant.externalId.origin);
+
+      const customClaims = buildJwtCustomClaims(
+        roles,
+        tenantId,
+        selfcareId,
+        tenant.externalId.origin,
+        tenant.externalId.value
+      );
+
+      const { serialized: sessionToken } = await sessionTokenGenerator.generate(
+        {
+          ...sessionClaims,
+          ...customClaims,
+        }
+      );
+      return sessionToken;
```

Comment:
I have a doubt here, in Scala we also return the `rateLimitStatus` which is not present in the openApi's response, is it an error in the Scala code? @galales

Reply:
I see rate limiting headers defined in the spec for this endpoint, but maybe I didn't understand your point.

cc @ecamellini this is another endpoint I forgot to mention where the rate limiting is applied in the middle of the business logic

Reply:
Yep, tracked here: https://pagopa.atlassian.net/browse/IMN-774